File: packages/typescript/service/tests/aggregation.test.ts
import { describe, it, expect } from "vitest";
import { AuditAggregator, AggregatedEntry } from "../src/aggregation.js";

function makeEvent(overrides: Partial<{
  actor_key: string;
  capability: string;
  failure_type: string;
  detail: string | null;
  timestamp: string;
}> = {}): Record<string, unknown> {
  return {
    actor_key: "agent-1",
    capability: "search_flights",
    failure_type: "scope_insufficient",
    detail: "Missing scope travel.search",
    timestamp: new Date("2026-01-01T00:00:30Z").toISOString(),
    ...overrides,
  };
}

describe("AuditAggregator", () => {
  it("single event not aggregated", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t = new Date("2026-01-01T00:00:30Z");
    agg.submit(makeEvent({ timestamp: t.toISOString() }));

    const results = agg.flush(new Date(t.getTime() + 61_000));
    expect(results).toHaveLength(1);
    expect(results[0]).not.toHaveProperty("count");
  });

  it("two identical events aggregated", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t1 = new Date("2026-01-01T00:00:10Z");
    const t2 = new Date("2026-01-01T00:00:20Z");
    agg.submit(makeEvent({ timestamp: t1.toISOString() }));
    agg.submit(makeEvent({ timestamp: t2.toISOString() }));

    const results = agg.flush(new Date(t1.getTime() + 61_000));
    expect(results).toHaveLength(1);
    const entry = results[0] as AggregatedEntry;
    expect(entry.count).toBe(2);
    expect(entry.event_class).toBe("repeated_low_value_denial");
    expect(entry.retention_tier).toBe("aggregate_only");
    expect(entry.first_seen).toBe(t1.toISOString());
    expect(entry.last_seen).toBe(t2.toISOString());
  });

  it("different keys not aggregated", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t = new Date("2026-01-01T00:00:10Z");
    agg.submit(makeEvent({ actor_key: "agent-1", timestamp: t.toISOString() }));
    agg.submit(makeEvent({ actor_key: "agent-2", timestamp: t.toISOString() }));

    const results = agg.flush(new Date(t.getTime() + 61_000));
    expect(results).toHaveLength(2);
    expect(results[0]).not.toHaveProperty("count");
    expect(results[1]).not.toHaveProperty("count");
  });

  it("representative_detail truncated to 200 chars", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t = new Date("2026-01-01T00:00:10Z");
    agg.submit(makeEvent({ detail: "x".repeat(500), timestamp: t.toISOString() }));
    agg.submit(makeEvent({ timestamp: new Date(t.getTime() + 5_000).toISOString() }));

    const results = agg.flush(new Date(t.getTime() + 61_000));
    const entry = results[0] as AggregatedEntry;
    expect(entry.representative_detail).toHaveLength(200);
  });

  it("representative_detail nullable", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t = new Date("2026-01-01T00:00:10Z");
    agg.submit(makeEvent({ detail: null, timestamp: t.toISOString() }));
    agg.submit(makeEvent({ detail: null, timestamp: new Date(t.getTime() + 5_000).toISOString() }));

    const results = agg.flush(new Date(t.getTime() + 61_000));
    const entry = results[0] as AggregatedEntry;
    expect(entry.representative_detail).toBeNull();
  });

  it("window boundary separates events", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t1 = new Date("2026-01-01T00:00:50Z");
    const t2 = new Date("2026-01-01T00:01:10Z"); // next window
    agg.submit(makeEvent({ timestamp: t1.toISOString() }));
    agg.submit(makeEvent({ timestamp: t2.toISOString() }));

    const r1 = agg.flush(new Date(t1.getTime() + 61_000));
    expect(r1).toHaveLength(1);
    expect(r1[0]).not.toHaveProperty("count");

    const r2 = agg.flush(new Date(t2.getTime() + 61_000));
    expect(r2).toHaveLength(1);
    expect(r2[0]).not.toHaveProperty("count");
  });

  it("pre-auth capability bucket", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t = new Date("2026-01-01T00:00:10Z");
    agg.submit(makeEvent({ capability: "_pre_auth", failure_type: "invalid_token", timestamp: t.toISOString() }));
    agg.submit(makeEvent({ capability: "_pre_auth", failure_type: "invalid_token", timestamp: new Date(t.getTime() + 5_000).toISOString() }));

    const results = agg.flush(new Date(t.getTime() + 61_000));
    expect(results).toHaveLength(1);
    const entry = results[0] as AggregatedEntry;
    expect(entry.grouping_key.capability).toBe("_pre_auth");
  });

  it("toAuditDict produces expected shape", () => {
    const agg = new AuditAggregator({ windowSeconds: 60 });
    const t = new Date("2026-01-01T00:00:10Z");
    agg.submit(makeEvent({ timestamp: t.toISOString() }));
    agg.submit(makeEvent({ timestamp: new Date(t.getTime() + 5_000).toISOString() }));

    const results = agg.flush(new Date(t.getTime() + 61_000));
    const entry = results[0] as AggregatedEntry;
    const d = entry.toAuditDict();
    expect(d.entry_type).toBe("aggregated");
    expect(d.event_class).toBe("repeated_low_value_denial");
    expect(d.retention_tier).toBe("aggregate_only");
    expect(d.count).toBe(2);
    expect(d.grouping_key).toBeDefined();
    expect(d.window).toBeDefined();
  });
});

File: packages/go/service/aggregation_test.go
package service

import (
	"testing"
	"time"
)

func TestAggregator_SingleEventPassesThrough(t *testing.T) {
	agg := NewAuditAggregator(60)
	now := time.Date(2026, 1, 1, 0, 0, 30, 0, time.UTC)

	agg.Submit(map[string]any{
		"actor_key":    "agent:test",
		"capability":   "search",
		"failure_type": "scope_insufficient",
		"timestamp":    now.Format(time.RFC3339),
	})

	flushTime := time.Date(2026, 1, 1, 0, 1, 1, 0, time.UTC)
	results := agg.Flush(flushTime)
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}
	entry, ok := results[0].(map[string]any)
	if !ok {
		t.Fatal("single event should pass through as map[string]any")
	}
	if entry["actor_key"] != "agent:test" {
		t.Error("entry should preserve original fields")
	}
}

func TestAggregator_MultipleEventsAggregate(t *testing.T) {
	agg := NewAuditAggregator(60)
	base := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)

	for i := 0; i < 5; i++ {
		agg.Submit(map[string]any{
			"actor_key":    "agent:spam",
			"capability":   "search",
			"failure_type": "scope_insufficient",
			"timestamp":    base.Add(time.Duration(i) * time.Second).Format(time.RFC3339),
			"detail":       "first detail",
		})
	}

	flushTime := base.Add(61 * time.Second)
	results := agg.Flush(flushTime)
	if len(results) != 1 {
		t.Fatalf("expected 1 aggregated result, got %d", len(results))
	}

	aggEntry, ok := results[0].(*AggregatedEntry)
	if !ok {
		t.Fatal("multiple events should produce *AggregatedEntry")
	}
	if aggEntry.Count != 5 {
		t.Errorf("count = %d, want 5", aggEntry.Count)
	}
	if aggEntry.EventClass != "repeated_low_value_denial" {
		t.Errorf("event_class = %q, want repeated_low_value_denial", aggEntry.EventClass)
	}
	if aggEntry.RetentionTier != "aggregate_only" {
		t.Errorf("retention_tier = %q, want aggregate_only", aggEntry.RetentionTier)
	}
}

func TestAggregator_DifferentKeysNotMerged(t *testing.T) {
	agg := NewAuditAggregator(60)
	base := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)

	agg.Submit(map[string]any{
		"actor_key": "agent:a", "capability": "search",
		"failure_type": "scope_insufficient", "timestamp": base.Format(time.RFC3339),
	})
	agg.Submit(map[string]any{
		"actor_key": "agent:b", "capability": "search",
		"failure_type": "scope_insufficient", "timestamp": base.Format(time.RFC3339),
	})

	results := agg.Flush(base.Add(61 * time.Second))
	if len(results) != 2 {
		t.Fatalf("different actor_keys should produce 2 results, got %d", len(results))
	}
}

func TestAggregator_FlushDoesNotEmitOpenWindows(t *testing.T) {
	agg := NewAuditAggregator(60)
	now := time.Date(2026, 1, 1, 0, 0, 30, 0, time.UTC)

	agg.Submit(map[string]any{
		"actor_key": "agent:test", "capability": "search",
		"failure_type": "scope_insufficient", "timestamp": now.Format(time.RFC3339),
	})

	results := agg.Flush(now.Add(10 * time.Second))
	if len(results) != 0 {
		t.Fatalf("should not flush open windows, got %d results", len(results))
	}
}

func TestAggregatedEntry_ToAuditDict(t *testing.T) {
	ae := &AggregatedEntry{
		EventClass:    "repeated_low_value_denial",
		RetentionTier: "aggregate_only",
		GroupingKey: map[string]string{
			"actor_key": "agent:test", "capability": "search", "failure_type": "scope_insufficient",
		},
		WindowStart:          "2026-01-01T00:00:00Z",
		WindowEnd:            "2026-01-01T00:01:00Z",
		Count:                10,
		FirstSeen:            "2026-01-01T00:00:05Z",
		LastSeen:             "2026-01-01T00:00:55Z",
		RepresentativeDetail: "Insufficient scope",
	}
	d := ae.ToAuditDict()
	if d["entry_type"] != "aggregated" {
		t.Errorf("entry_type = %q, want aggregated", d["entry_type"])
	}
	if d["count"] != 10 {
		t.Errorf("count = %v, want 10", d["count"])
	}
	if d["capability"] != "search" {
		t.Errorf("capability = %v, want search", d["capability"])
	}
}
